enhancements/etcd/protecting-etcd-quorum-during-control-plane-scaling.md

```diff
@@ -0,0 +1,450 @@
+is preserved throughout the operation.
+
+This process is not sustainable and we must provide safety mechanisms on top of the existing etcd quorum guard to make
+this process both easier and safer.
+
+### Goals
+
+* Provide the etcd Operator with the ability to control when a Control Plane Machine is removed from the cluster
+* Allow the etcd Operator to prevent removal of etcd members until a replacement member has been promoted to a voting
+  member
+* Allow the etcd Operator to remove an etcd member from the etcd cluster before the Machine is terminated to prevent a
+  degraded etcd cluster
+* Allow an escape hatch from the protection when surging the capacity with new Control Plane Machines is unavailable
+  (for example in metal environments with limited capacity)
+
+
+### Non-Goals
+
+* Automation of scaling operations on Machines
+* Horizontal scaling of the etcd cluster and Control Plane
+* Providing these protection mechanisms when the Machine API is [not functional](#When-is-Machine-API-Functional)
+* Recovering unhealthy etcd clusters
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator of a managed OpenShift cluster, I want to be able to automate the scaling operations of Control Plane
+Machines without having to manually ensure the safety of the etcd cluster.
+
+#### Story 2
+
+As a developer of OpenShift, I want to implement safety mechanisms that adhere to best practices for etcd scaling
+operations to protect end users from potential quorum losses and data losses.
+
+#### Story 3
+
+As an end user of OpenShift, I want to be able to increase the size of my Control Plane without having to know the
+intricacies of etcd and protecting its quorum.
+
+### API Extensions
+
+This enhancement does not introduce any new API extensions.
+
+### Implementation Details/Notes/Constraints
+
+#### Requirements for etcd safety
+
+* The number of voting members should equal the desired number of Control Plane Machines (and this should be an odd
+  number)
+  * We will only deviate from this to add a replacement member
+  * Once the new member is added, we should remove old members as soon as possible to reduce the risk of degrading the
+    cluster while having an even number of voting members
+* Existing voting members should not be removed until their replacements have been promoted to voting members
+  * By starting new etcd members as [learners](https://etcd.io/docs/v3.3/learning/learner/#raft-learner), we can ensure
+    that the new member is fully "caught-up" and promotable to a full voting member before we start the removal process
+    of the old member
+  * This protects etcd from potential inconsistencies in its data if the cluster were to have some interruption shortly
+    after a new member joins the cluster
+  * This also ensures that we keep the full etcd data on disk on a minimum of the desired number of Control Plane
+    Machines at all times
+
+#### Protecting etcd during scaling operations
```

Comment:
do we need to protect from an accidental etcd rollout? (like somebody change etcd config or hit upgrade button during this surgery) ?

Reply:
This feels like it could be tricky to co-ordinate, because the operation could be either a single machine being replaced or many.

If the etcd operator blocked rolling out config changes while a Machine was marked as deleted, that could be safe. 

We expect that when a new machine is being added, the old one is deleted signalling the etcd member that will eventually be removed. In the case of removing multiple machines, we have designed the mechanism so that we can delete multiple machines at once, and the mechanism will roll out the new etcd members 1 by 1, so this should also be safe.

Not sure if it's strictly necessary though, not sure what effect changing the config can have to the etcd cluster, I assume we restrict the configuration changes that end users can make for the etcd instances? If we are in the learning phase, a config change sounds like it shouldn't be too detrimental?

Reply:
From my understanding users really can't change the etcd config very much (other than update what ciphers are used) since the etcd-operator controls the env vars that are rolled out to the etcd static pods:
https://github.com/openshift/cluster-etcd-operator/blob/d8c023e4d217aae762b99322923c68101c4b3ffd/pkg/operator/targetconfigcontroller/targetconfigcontroller.go#L194-L203
https://github.com/openshift/cluster-etcd-operator/blob/d8c023e4d217aae762b99322923c68101c4b3ffd/pkg/etcdenvvar/etcd_env.go#L71-L84

They can however force a redeployment of all the etcd static-pods and trigger a roll-out to the next revision (even if it's without any changes) by updating the etcd/cluster CR e.g:
```sh
oc patch etcd cluster -p='{"spec": {"forceRedeploymentReason": "foo"}}' --type=merge
```

All that will do (via the revision controller) is essentially restart the member pod on each serially.

Since that rollout doesn't change the cluster membership I don't think it would impact an ongoing scaling operation.
At worst it might delay the scaling by a little bit if the operator is attempting to promote a learner member who's pod is being recreated with the new revision.

The scaling operation itself would actually cause multiple revision rollouts as it updates the cluster membership when adding/removing members. 
@mfojtik So that should be okay unless you had another case in mind.

Reply:
> or hit upgrade button during this surgery

Upgrades should be blocked via the quorum guard during an ongoing scaling operation. That's currently [outlined](https://github.com/openshift/enhancements/pull/943/files#diff-8f08f89622ea85c6276166feaa7be109adaae9cf5d280b18ef20f4f218fce87cR193-R197) in the proposal:
```
While a scaling operation is occurring, the etcd quorum guard will prevent the draining of any of the Control Plane
Machines, until the new etcd member has been promoted and the old etcd member removed from the cluster.
This in turn means that updates caused during upgrades (for example changes to MachineConfig) will be blocked while the
scaling operation occurs.
This will delay the upgrade process, but should not block it indefinitely unless an issue occurs.
```